File: PKHImageCache/PKHImageDownloadOperation.m
//
//  PKHImageDownloadOperation.m
//  PKHImageCache
//
//  Created by Patrick Hanlon on 5/4/15.
//  Copyright (c) 2015 Patrick Hanlon. All rights reserved.
//

#import "PKHImageDownloadOperation.h"
#import "PKHImageCache.h"

@interface PKHImageDownloadOperation () <NSURLSessionDelegate, NSURLSessionDownloadDelegate>
{
    NSURLSession *_session;
    NSURLSessionDownloadTask *_downloadTask;
    PKHImageDownloaderCompletionBlock _completionBlock;
}

@end

@implementation PKHImageDownloadOperation

#pragma mark - Public

- (void)startWithCompletion:(PKHImageDownloaderCompletionBlock)completionBlock
{
    _completionBlock = [completionBlock copy];
    _downloadTask = [self.session downloadTaskWithURL:[NSURL URLWithString:self.imageURLString]];
    [_downloadTask resume];
}

#pragma mark - Private

- (NSURLSession *)session
{
    if (!_session) {
        NSURLSessionConfiguration *sessionConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
        _session = [NSURLSession sessionWithConfiguration:sessionConfig delegate:self delegateQueue:nil];
    }
    return _session;
}

#pragma mark - NSURLSessionDownloadDelegate methods

- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location
{
    NSData *data = [NSData dataWithContentsOfURL:location];
    UIImage *image = [UIImage imageWithData:data];
    
    if (_completionBlock != nil) {
        _completionBlock(image, [NSURL URLWithString:self.imageURLString]);
    }

    [_session finishTasksAndInvalidate];
}


@end

File: PKHImageCache/PKHImageDownloadOperation.swift
//
//  PKHImageDownloadOperation.swift
//  PKHIC-Example
//
//  Created by Patrick Hanlon on 6/24/15.
//  Copyright (c) 2015 pkh. All rights reserved.
//

import Foundation
import UIKit

typealias PKHImageDownloaderCompletionBlock = (image: UIImage, imageURL: NSURL?) -> Void

class PKHImageDownloadOperation: NSObject, NSURLSessionDownloadDelegate {
    
    var imageURLString: String!
    var downloadTask: NSURLSessionDownloadTask!
    var completionBlock: PKHImageDownloaderCompletionBlock!
    
    lazy var session: NSURLSession = {
        let sessionConfig: NSURLSessionConfiguration = NSURLSessionConfiguration.defaultSessionConfiguration()
        var newSession: NSURLSession = NSURLSession(configuration: sessionConfig, delegate: self, delegateQueue: nil)
        return newSession
    }()
    
    // MARK: Public Interface
    
    func start(imageURLString: String, completionBlock: PKHImageDownloaderCompletionBlock) {
        self.imageURLString = imageURLString
        self.completionBlock = completionBlock
        self.downloadTask = self.session.downloadTaskWithURL(NSURL(string: self.imageURLString)!)
        self.downloadTask.resume()
    }
    
    // MARK: NSURLSessionDownloadDelegate Methods
    
    func URLSession(session: NSURLSession, downloadTask: NSURLSessionDownloadTask, didFinishDownloadingToURL location: NSURL) {
        
        let data: NSData = NSData(contentsOfURL: location)!
        let image: UIImage = UIImage(data: data)!
        
        if let completion = self.completionBlock {
            completion(image: image, imageURL: NSURL(string: self.imageURLString!))
        }
        
        session.finishTasksAndInvalidate()
    }
    
}
